src/syscheckd/src/whodata/win_whodata.c

```diff
@@ -555,9 +555,18 @@ int restore_audit_policies() {
         merror(FIM_ERROR_WHODATA_RESTORE_POLICIES);
         return 1;
     }
+
     // Get the current policies
     char *cmd_output = NULL;
-    const int wm_exec_ret_code = wm_exec(command, &cmd_output, &result_code, 5, NULL);
+    int wm_exec_ret_code;
+    int i, retries = 5;
+
+    for (i = 0; i <= retries; i++) {
```

Comment:
If we do it this way, I would at least include the current retry number and the max retries value in the logs to address which attempts and how many attempts are being made.

Reply:
Very interesting yes. I also propose to add some verbose on the loop to show the number of retries, and in case of not having been able to perform it successfully, to clarify it at the end:

```suggestion
    // Get the current policies
    char *cmd_output = NULL;
    int wm_exec_ret_code;
    int retries = 5;
    BOOL cmd_failed = 0;

    for (int i = 0; i <= retries; i++) {
        wm_exec_ret_code = wm_exec(command, &cmd_output, &result_code, 5+i, NULL);

        if (wm_exec_ret_code < 0) {
            merror(FIM_ERROR_WHODATA_AUDITPOL, "failed to execute command");
            cmd_failed = 1;
        } else if (wm_exec_ret_code == 1) {
            merror(FIM_ERROR_WHODATA_AUDITPOL, "time overtaken while running the command");
            os_free(cmd_output);
            cmd_failed = 1;
        } else if (!wm_exec_ret_code && result_code) {
            char error_msg[OS_MAXSTR];
            snprintf(error_msg, OS_MAXSTR, FIM_ERROR_WHODATA_AUDITPOL, "command returned failure'. Output: '%s");
            merror(error_msg, cmd_output);
            os_free(cmd_output);
            cmd_failed = 1;
        }

        if (!cmd_failed) {
            break;
        }
                
        if (i == retries) {
            merror("After X attempts the Auditpol command could not be executed successfully.");
        } else {
            merror("Auditpol command failed, attempt number X");
        }
    }
```

Reply:
Hi team!
I think the second one is an exciting proposal, but I would modify it as follows:
```suggestion
    // Get the current policies
    char *cmd_output = NULL;
    int wm_exec_ret_code;
    int retries = 5;
    BOOL cmd_failed = 0;
    for (int i = 0; i <= retries; i++) {
        wm_exec_ret_code = wm_exec(command, &cmd_output, &result_code, 5+i, NULL);
        if (wm_exec_ret_code < 0) {
            merror(FIM_ERROR_WHODATA_AUDITPOL, "failed to execute command");
            cmd_failed = 1;
        } else if (wm_exec_ret_code == 1) {
            merror(FIM_ERROR_WHODATA_AUDITPOL, "time overtaken while running the command");
            os_free(cmd_output);
            cmd_failed = 1;
        } else if (!wm_exec_ret_code && result_code) {
            char error_msg[OS_MAXSTR];
            snprintf(error_msg, OS_MAXSTR, FIM_ERROR_WHODATA_AUDITPOL, "command returned failure'. Output: '%s");
            merror(error_msg, cmd_output);
            os_free(cmd_output);
            cmd_failed = 1;
        }
        if (!cmd_failed) {
            break;
        } else {
            merror("Auditpol command failed, attempt number X");
        }
               
    }
    if (i == retries) {
       merror("After X attempts the Auditpol command could not be executed successfully.");
    } 
```

This way, it is not necessary to compare in every cycle loop if `(i == retries)` which will be almost always false.

Reply:
**Warning:** `i` will be out of scope. However, if you move the variable into the scope, the value here should not be compared with retries. The desired scenario is: `i = 6` and  `retries = 5`, triggering an error log. Keep in mind that if `i = 5`, this case could be considered successful.